### PR TITLE
Transfer learning: RL training using loaded reward model

### PR DESCRIPTION
**Pull request by AdamGleave**
This PR creates a new registry for reward function loaders, and adds support to `expert_demos` to load any reward function supported in this registry, wrapping the environment.

Specifically, we actually create two registries: one for `RewardNet` objects, and one for callables which take obs-act-obs triples and return rewards. The latter is a more general interface, and is what is needed by `expert_demos`. The former, `RewardNet`, is more constrained (it would exclude e.g. `DiscrimNetGAIL`) but could be useful in cases where one needs a TensorFlow reward model (e.g. for fine-tuning/further training).

### CODE REVIEW

File: src/imitation/rewards/serialize.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    "BasicShapedRewardNet": reward_net.BasicShapedRewardNet,
+}
+_add_reward_net_loaders(REWARD_NETS)
+
+
+def _load_discrim_net(cls):
+  def f(path: str, venv: VecEnv):
+    discrim_net = cls.load(path)
+    # TODO(gleave): expose train reward as well? (hard due to action probs?)
+    return discrim_net.reward_test
```
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/rewards/serialize.py]: Maybe Discrim net builds its own sess

File: src/imitation/rewards/serialize.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+})
+
+
+def load_reward_net_as_fn(path: str, env: VecEnv) -> RewardFn:
+  reward_type, shaped, reward_path = path.split(':')
+  reward_net_loader = reward_net_registry.get(reward_type)
+  assert shaped in ["True", "False"]
+  shaped = shaped == "True"
+
+  # TODO(adam): leaks session
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py]: I'd be OK with this, but it feels just like punting the problem elsewhere. At the current callsite in scripts/expert_demos.py:105 it seems like most the code needs the reward model. Probably we'd want to deallocate the session when the `RewardVecEnvWrapper` gets `close`(d)?

File: src/imitation/rewards/serialize.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py]: I think it actually should have session creation logic too, added. (It worked because it used the default session, but in general this can cause name clashes.)

File: src/imitation/rewards/serialize.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+
+_add_discrim_net_loaders({
+    "DiscrimNetAIRL": discrim_net.DiscrimNetAIRL,
+    "DiscrimNetGAIL": discrim_net.DiscrimNetGAIL,
+})
+
+
+def load_reward_net_as_fn(path: str, env: VecEnv) -> RewardFn:
+  reward_type, shaped, reward_path = path.split(':')
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py]: I don't want to add a `reward_shaped` parameter to Sacred since many reward models won't have a shaped/unshaped variant. This is just specific to `RewardNet`, and really just to AIRL.

I've changed this to make the `reward_type` (top-level, in `load_reward`) be either `RewardNet_unshaped` or `RewardNet_shaped`, and then pass a `shaped` term to `load_reward_net_as_fn`.

So the end user would specify e.g. `with reward_type=RewardNet_unshaped reward_path=BasicShapedRewardNet:~/output/train/path/to/reward_net`

File: src/imitation/rewards/serialize.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+reward_registry.register(key='zero', value=load_zero)
+reward_registry.register(key="RewardNet", value=load_reward_net_as_fn)
+
+
+def load_reward(reward_type: str, reward_path: str,
+                venv: VecEnv) -> RewardFn:
+  """Load serialized policy.
+
+  Args:
+    reward_type: A key in `reward_registry`, e.g. `RewardNet`.
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py]: The set of valid types is going to depend on what third-party code is loaded, since they can also register things, but it seems helpful to include the defaults in the doc; added.

**Comment by AdamGleave** [comment]
Python type annotations for `Callable` do not support keyword arguments annoyingly, and `pytype` also has trouble with `if-else` branches leading to callables of different types, so I've: (a) made the `steps` argument mandatory (and adjusted existing reward functions to ignore it) and (b) made it a positional argument not a keyword argument.

File: src/imitation/rewards/serialize.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/rewards/serialize.py]: nice decorator `@utils.docstring_parameter`

File: src/imitation/rewards/serialize.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/rewards/serialize.py]: That seems ok 👌

File: src/imitation/rewards/serialize.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/rewards/serialize.py]: Agree that `reward_fn` and its Session should be closed once `venv = RewardVecEnvWrapper(venv, reward_fn)` is no longer used (nearly at the end of the function).

[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/rewards/serialize.py]: (`RewardVecWrapper` wouldn't make a good contextmanager because it isn't guaranteed to hold a `Session` (and wouldn't have access to reward_fn's Session anyways).)

[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/rewards/serialize.py]: Conditionally setting up a context from `reward_fn` seems hard to do via a vanilla context + a no-op context switch.

https://github.com/HumanCompatibleAI/imitation/blob/2d1f29fb636b3891e8ce0c7c3845976f2df514fd/src/imitation/scripts/expert_demos.py#L99-L110

[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/rewards/serialize.py]: So this might be a good use case for contextlib.ExitStack.

Something like
```
with ExitStack() as exit_stack:
  ...
  if reward_type is not None:
    reward_fn, resources = load_reward(...)   # type: Tuple[RewardFn, List["Implements .close()"]]
    for resource in resources: 
      exit_stack.push(resource)
    venv = RewardVecEnvWrapper(venv, reward_fn)
```

[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/rewards/serialize.py]: Thanks ! I'd never seen `ExitStack`. I'm going to address this in a separate PR since I want to change things for `policies/serialize.py` as well.
